subsys/net/ip/ipv4.c

```diff
@@ -403,21 +403,27 @@ enum net_verdict net_ipv4_input(struct net_pkt *pkt, bool is_loopback)
 		}
 		break;
 
-#if defined(CONFIG_NET_L2_VIRTUAL)
+#if defined(CONFIG_NET_L2_IPIP)
 	case IPPROTO_IPV6:
 	case IPPROTO_IPIP: {
-		struct net_addr remote_addr;
+		struct sockaddr_in remote_addr = { 0 };
+		struct net_if *tunnel_iface;
 
-		remote_addr.family = AF_INET;
-		net_ipv4_addr_copy_raw((uint8_t *)&remote_addr.in_addr, hdr->src);
+		remote_addr.sin_family = AF_INET;
+		net_ipv4_addr_copy_raw((uint8_t *)&remote_addr.sin_addr, hdr->src);
+
+		net_pkt_set_remote_address(pkt, (struct sockaddr *)&remote_addr,
+					   sizeof(struct sockaddr_in));
 
 		/* Get rid of the old IP header */
 		net_pkt_cursor_restore(pkt, &hdr_start);
 		net_pkt_pull(pkt, net_pkt_ip_hdr_len(pkt) +
 			     net_pkt_ipv4_opts_len(pkt));
 
-		return net_virtual_input(net_pkt_iface(pkt), &remote_addr,
-					 pkt);
+		tunnel_iface = net_ipip_get_virtual_interface(net_pkt_iface(pkt));
+		if (tunnel_iface != NULL && net_if_l2(tunnel_iface)->recv != NULL) {
+			return net_if_l2(tunnel_iface)->recv(net_pkt_iface(pkt), pkt);
+		}
```

Comment:
Is the behavior change intended here? If `tunnel_iface` is `NULL` or doesn't have a `recv` function, should we continue or return?
I think it's safe, as the `verdict` is initialized to `NET_DROP`, just making sure.

Reply:
Yes, the purpose was that the function will bail out few lines below.